File: Chapter 8/8.27-multiviewport/ConfigureEnvironment/GLCoreProfileView.m
//
//  GLCoreProfileView.m
//  ConfigureEnvironment
//
//  Created by 陈杰 on 26/10/2017.
//  Copyright © 2017 陈杰. All rights reserved.
//

#import "GLCoreProfileView.h"
#import <OpenGL/OpenGL.h>
#import <GLKit/GLKit.h>
#import "TextureManager.h"
#import "TDModelManger.h"

@interface GLCoreProfileView()

@property (nonatomic, strong) NSTimer *lifeTimer;
@property (nonatomic, assign) CGFloat lifeDuration;
@property (nonatomic, assign) BOOL paused;

@property (atomic, assign) GLuint program;
@property (atomic, assign) GLuint vertexArray;
@property (atomic, assign) GLuint position_buffer;
@property (atomic, assign) GLuint index_buffer;
@property (atomic, assign) GLuint uniform_buffer;

@property (atomic, assign) GLint transformBlockIndex;
@property (atomic, assign) GLint transformBlockLocation;

@end

@implementation GLCoreProfileView
#pragma mark - lifecycle methods
- (instancetype)initWithCoder:(NSCoder *)decoder {
    NSOpenGLPixelFormatAttribute pixelFormatAttributes[] = {
        NSOpenGLPFAColorSize, 32,
        NSOpenGLPFADepthSize, 24,
        NSOpenGLPFAStencilSize, 8,
        NSOpenGLPFAAccelerated,
        NSOpenGLPFAOpenGLProfile, NSOpenGLProfileVersion4_1Core,
        0
    };
    NSOpenGLPixelFormat *pixelFormat = [[NSOpenGLPixelFormat alloc] initWithAttributes:pixelFormatAttributes];
    NSOpenGLContext *openGLContext = [[NSOpenGLContext alloc] initWithFormat:pixelFormat shareContext:nil];
    
    
    if (self = [super initWithCoder:decoder]) {
        [self setOpenGLContext:openGLContext];
        [self.openGLContext makeCurrentContext];
        GLint sync = 1;
        CGLSetParameter(CGLGetCurrentContext(), kCGLCPSwapInterval, &sync);
        
        _lifeTimer = [NSTimer scheduledTimerWithTimeInterval:0.02 target:self selector:@selector(lifeTimerUpdate) userInfo:nil repeats:YES];
        _paused = NO;
    
        __weak __typeof(self) weakself = self;
        [NSEvent addLocalMonitorForEventsMatchingMask:NSEventMaskKeyDown handler:^NSEvent * _Nullable(NSEvent * _Nonnull aEvent) {
            [weakself keyDown:aEvent];
            return aEvent;
        }];
    }
    return self;
}

- (void)keyDown:(NSEvent *)event {
    if ([event.characters isEqualToString:@"f"]) {
        
    } else if ([event.characters isEqualToString:@"d"]) {

    } else if ([event.characters isEqualToString:@"w"]) {
    } else if ([event.characters isEqualToString:@"p"]) {
        self.paused = !self.paused;
    } else if ([event.characters isEqualToString:@"+"]) {

    } else if ([event.characters isEqualToString:@"-"]) {

    }
    NSLog(@"%@",event.characters);
}

- (void)dealloc {
    [_lifeTimer invalidate];
    _lifeTimer = nil;

//    glDeleteBuffers(1, &_patchBuffer);
//    glDeleteBuffers(1, &_cageIndicesBuffer);
//    glDeleteVertexArrays(1, &_vertexArray);
//    glDeleteProgram(_tessProgram);
//    glDeleteProgram(_deawCpProgram);
}

- (void)prepareOpenGL {
    NSLog(@"Version: %s", glGetString(GL_VERSION));
    NSLog(@"Renderer: %s", glGetString(GL_RENDERER));
    NSLog(@"Vendor: %s", glGetString(GL_VENDOR));
    NSLog(@"GLSL Version: %s", glGetString(GL_SHADING_LANGUAGE_VERSION));
    
    [self loadShaders];

    glGenVertexArrays(1, &_vertexArray);
    glBindVertexArray(_vertexArray);
    
    static const GLushort vertex_indices[] = {
        0, 1, 2,
        2, 1, 3,
        2, 3, 4,
        4, 3, 5,
        4, 5, 6,
        6, 5, 7,
        6, 7, 0,
        0, 7, 1,
        6, 0, 2,
        2, 4, 6,
        7, 5, 3,
        7, 3, 1
    };
    
    static const GLfloat vertex_positions[] = {
        -0.25f, -0.25f, -0.25f,
        -0.25f,  0.25f, -0.25f,
        0.25f, -0.25f, -0.25f,
        0.25f,  0.25f, -0.25f,
        0.25f, -0.25f,  0.25f,
        0.25f,  0.25f,  0.25f,
        -0.25f, -0.25f,  0.25f,
        -0.25f,  0.25f,  0.25f,
    };
    
    glGenBuffers(1, &_position_buffer);
    glBindBuffer(GL_ARRAY_BUFFER, _position_buffer);
    glBufferData(GL_ARRAY_BUFFER, sizeof(vertex_positions), vertex_positions, GL_STATIC_DRAW);
    glVertexAttribPointer(0, 3, GL_FLOAT, GL_FALSE, 0, NULL);
    glEnableVertexAttribArray(0);
    
    glGenBuffers(1, &_index_buffer);
    glBindBuffer(GL_ELEMENT_ARRAY_BUFFER, _index_buffer);
    glBufferData(GL_ELEMENT_ARRAY_BUFFER, sizeof(vertex_indices), vertex_indices, GL_STATIC_DRAW);
    
    glGenBuffers(1, &_uniform_buffer);
    glBindBuffer(GL_UNIFORM_BUFFER, _uniform_buffer);
    glBufferData(GL_UNIFORM_BUFFER, 4 * sizeof(GLKMatrix4), NULL, GL_DYNAMIC_DRAW);
    
    glEnable(GL_CULL_FACE);
    // glFrontFace(GL_CW);
    glEnable(GL_DEPTH_TEST);
    glDepthFunc(GL_LEQUAL);
}

- (void)reshape {
//    NSRect bounds = [self bounds];
//    glViewport(0, 0, NSWidth(bounds), NSHeight(bounds));
    [self updateViewPorts];
}

- (void)drawRect:(NSRect)dirtyRect {
    static const GLfloat black[] = { 0.0f, 0.0f, 0.0f, 1.0f };
    glClearBufferfv(GL_COLOR, 0, black);
    static const GLfloat one = 1.0f;
    glClearBufferfv(GL_DEPTH, 0, &one);
    
    [self updateViewPorts];
    
    glBindBufferBase(GL_UNIFORM_BUFFER, 0, _uniform_buffer);
    NSRect bounds = [self bounds];
    GLKMatrix4 projectionMatrix = GLKMatrix4MakePerspective(GLKMathDegreesToRadians(50.0f), NSWidth(bounds)/NSHeight(bounds), 0.1f, 1000.0f);
    GLKMatrix4 *mv_matrix_array = (GLKMatrix4 *)glMapBufferRange(GL_UNIFORM_BUFFER, 0, 4 * sizeof(GLKMatrix4), GL_MAP_WRITE_BIT | GL_MAP_INVALIDATE_BUFFER_BIT);
    float f = _lifeDuration * 0.3;
    
    for (int i = 0; i < 4; i++) {
        GLKMatrix4 rotateX = GLKMatrix4MakeRotation(GLKMathDegreesToRadians((float)f * 81.0f * (float)(i + 1)), 1.0f, 0.0f, 0.0f);
        GLKMatrix4 rotateY = GLKMatrix4MakeRotation(GLKMathDegreesToRadians((float)f * 45.0f * (float)(i + 1)), 0.0f, 1.0f, 0.0f);
        GLKMatrix4 translation = GLKMatrix4MakeTranslation(0.0f, 0.0f, -1.0f);
        GLKMatrix4 mvMatrix = GLKMatrix4Multiply(translation, GLKMatrix4Multiply(rotateY, rotateX));
        GLKMatrix4 mvpMatrix = GLKMatrix4Multiply(projectionMatrix, mvMatrix);
        mv_matrix_array[i] = GLKMatrix4MakeWithArray(mvpMatrix.m);
    }
    glUnmapBuffer(GL_UNIFORM_BUFFER);
    
    glUseProgram(_program);
    glDrawElements(GL_TRIANGLES, 36, GL_UNSIGNED_SHORT, 0);
    glFlush();
}

#pragma mark - public methods

#pragma mark - private methods
- (BOOL)loadShaders {
    GLuint vertex_lineShader, ge_lineShader, frag_lineShader;
    NSString *vs_linePath = [[NSBundle mainBundle] pathForResource:@"linesadjacencyVS" ofType:@"glsl"];
    NSString *ge_linePath = [[NSBundle mainBundle] pathForResource:@"linesadjacencyGS" ofType:@"glsl"];
    NSString *fs_linePath = [[NSBundle mainBundle] pathForResource:@"linesadjacencyFS" ofType:@"glsl"];
    
    if (![self compileShader:&vertex_lineShader type:GL_VERTEX_SHADER filePath:vs_linePath]) {
        NSLog(@"Failed to compile GL_VERTEX_SHADER");
    }
    if (![self compileShader:&ge_lineShader type:GL_GEOMETRY_SHADER filePath:ge_linePath]) {
        NSLog(@"Failed to compile GL_TESS_CONTROL_SHADER");
    }
    if (![self compileShader:&frag_lineShader type:GL_FRAGMENT_SHADER filePath:fs_linePath]) {
        NSLog(@"Failed to compile GL_FRAGMENT_SHADER");
    }
    
    _program = glCreateProgram();
    glAttachShader(_program, vertex_lineShader);
    glAttachShader(_program, ge_lineShader);
    glAttachShader(_program, frag_lineShader);
    
    GLuint shaders2[] = {vertex_lineShader, ge_lineShader, frag_lineShader};
    for (int i = 0; i < 3; i++) {
        GLuint deleteShader = shaders2[i];
        glDeleteShader(deleteShader);
        deleteShader = 0;
    }
    
    if (![self linkProgram:_program]) {
        NSLog(@"Failed to link _tessProgram");
        if (_program != 0) {
            glDeleteProgram(_program);
            _program = 0;
        }
        return NO;
    }
    _transformBlockLocation = glGetUniformLocation(_program, "transform_block");
    _transformBlockIndex = glGetUniformBlockIndex(_program, "transform_block");
    glUniformBlockBinding(_program, _transformBlockIndex, 0);
    return YES;
}

- (BOOL)compileShader:(GLuint *)shader type:(GLenum)type filePath:(NSString *)path {
    const GLchar *shaderSource = [NSString stringWithContentsOfFile:path encoding:NSUTF8StringEncoding error:nil].UTF8String;
    *shader = glCreateShader(type);
    glShaderSource(*shader, 1, &shaderSource, nil);
    glCompileShader(*shader);
    
    GLint status = 0;
    glGetShaderiv(*shader, GL_COMPILE_STATUS, &status);
    if (status == 0) {
        GLint logLen = 0;
        glGetShaderiv(*shader, GL_INFO_LOG_LENGTH, &logLen);
        GLchar *infoLog = malloc(sizeof(char) * logLen);
        glGetShaderInfoLog(*shader, logLen, NULL, infoLog);
        NSLog(@"Shader at: %@", path);
        fprintf(stderr, "Info Log: %s\n", infoLog);
        
        glDeleteShader(*shader);
        return NO;
    }
    return YES;
}

- (BOOL)linkProgram:(GLuint)program {
    glLinkProgram(program);
    GLint status = 0;
    glGetProgramiv(program, GL_LINK_STATUS, &status);
    if (status == 0) {
        GLint logLen = 0;
        glGetProgramiv(program, GL_INFO_LOG_LENGTH, &logLen);
        GLchar *infoLog = malloc(sizeof(char) * logLen);
        glGetProgramInfoLog(program, logLen, NULL, infoLog);
        fprintf(stderr, "Prog Info Log: %s\n", infoLog);
        return NO;
    }
    return YES;
}

- (void)updateViewPorts {
    NSRect bounds = [self bounds];
    // Each rectangle will be 8/16 of the screen
    float viewport_width = (float)(8 * NSWidth(bounds)) / 16.0f;
    float viewport_height = (float)(8 * NSHeight(bounds)) / 16.0f;
    // Four rectangles - lower left first...
    glViewportIndexedf(0, 0, 0, viewport_width, viewport_height);
    // Lower right...
    glViewportIndexedf(1, NSWidth(bounds) - viewport_width, 0, viewport_width, viewport_height);
    // Upper left...
    glViewportIndexedf(2, 0, NSHeight(bounds) - viewport_height, viewport_width, viewport_height);
    // Upper right...
    glViewportIndexedf(3, NSWidth(bounds) - viewport_width, NSHeight(bounds) - viewport_height, viewport_width, viewport_height);
}

#pragma mark - listening methods
- (void)lifeTimerUpdate {
    _lifeDuration += _lifeTimer.timeInterval;
    // _lifeDuration为程序运行时间
    [self drawRect:self.bounds];
}

#pragma mark - accessor methods

@end

File: Chapter 12/PhongLighting/Source/GLCoreProfileView.swift
//
//  GLCoreProfileView.swift
//  PhongLighting
//
//  Created by chenjie on 2019/11/8.
//  Copyright © 2019 chenjie. All rights reserved.
//

import GLKit

struct UniformsBlock {
    var viewMatrix: GLKMatrix4
    var modelViewMatrix: GLKMatrix4
    var projectionMatrix: GLKMatrix4
}

struct UniformLocation {
    var ambientColorLocation: GLint
    var diffuseAlbedoLocation: GLint
    var specularAlbedoLocation: GLint
    var specularPowerLocation: GLint
}

class GLCoreProfileView: NSOpenGLView {
    //MARK:- Public properties
    // 除去暂停后实际渲染的时间
    var renderDuration: TimeInterval = 0
    var per_vertex = false
    var many_objects = false

    //MARK:- Private Properties    
    private var perFragmentProgram: GLuint = 0
    private var perVertexProgram: GLuint = 0
    private var uniformsBuffer: GLuint = 0
    private var perfragmentUniformsLocation: UniformLocation!
    private var perVertexUnifromsLocation: UniformLocation!
    
    //MARK:- Life Cycles
    override init?(frame frameRect: NSRect, pixelFormat format: NSOpenGLPixelFormat?) {
        super.init(frame: frameRect, pixelFormat: format)
        
        self.prepareOpenGLContex()
    }
    
    required init?(coder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }
        
    //MARK:- Private Methods: Layout
    override func reshape() {
        super.reshape()
        
        let bounds = self.bounds
        glViewport(0, 0, GLsizei(NSWidth(bounds)), GLsizei(NSHeight(bounds)))
    }
    
    //MARK:- Private Methods: Render
    private func prepareOpenGLContex() {
        let pixelFormatAttributes = [UInt32(NSOpenGLPFAColorSize), 32,
                                     UInt32(NSOpenGLPFADepthSize), 24,
                                     UInt32(NSOpenGLPFAStencilSize), 8,
                                     UInt32(NSOpenGLPFAAccelerated),
                                     UInt32(NSOpenGLPFAOpenGLProfile), UInt32(NSOpenGLProfileVersion4_1Core),
                                     0]
        guard let pixelFormat = NSOpenGLPixelFormat(attributes: pixelFormatAttributes) else { return }
        let currentOpenGLContext = NSOpenGLContext(format: pixelFormat, share: nil)
        self.openGLContext = currentOpenGLContext
        self.openGLContext?.makeCurrentContext()
        // 设置缓存交换频率和屏幕刷新同步
        var swapInterval = GLint(1)
        CGLSetParameter(CGLGetCurrentContext()!, kCGLCPSwapInterval, &swapInterval)
    }
    
    override func prepareOpenGL() {
        super.prepareOpenGL()
        print("Version: \(String(cString: glGetString(uint32(GL_VERSION))))")
        print("Renderer: \(String(cString: glGetString(uint32(GL_RENDERER))))")
        print("Vendor: \(String(cString: glGetString(GLenum(GL_VENDOR))))")
        print("GLSL Version: \(String(cString: glGetString(GLenum(GL_SHADING_LANGUAGE_VERSION))))")
        
        // 1. 准备OpenGL程序
        if !self.preparePerVertexProgram() {
            return
        }
        if !self.preparePerFragmentProgram() {
            return
        }
                
        // 2. 为统一变量准备缓存
        glGenBuffers(1, &self.uniformsBuffer)
        glBindBufferBase(GLenum(GL_UNIFORM_BUFFER), 0, self.uniformsBuffer)
        glBufferData(GLenum(GL_UNIFORM_BUFFER), MemoryLayout<UniformsBlock>.stride, nil, GLenum(GL_DYNAMIC_DRAW))
        glBindBufferBase(GLenum(GL_UNIFORM_BUFFER), 0, 0)

        // 3. 加载模型
        let successed = ModelManager.shareManager.loadObject(fileName: "sphere.sbm")
        if !successed {
            return
        }
        
        // 4. 开启面剔除和深度检测
        glEnable(GLenum(GL_CULL_FACE))
        glEnable(GLenum(GL_DEPTH_TEST))
        glDepthFunc(GLenum(GL_LEQUAL))
    }
    
    private func preparePerFragmentProgram() -> Bool {
        let shaders = ["perFragmentVS" : GLenum(GL_VERTEX_SHADER), "perFragmentFS" : GLenum(GL_FRAGMENT_SHADER)]
        let successed = self.prepareGLProgram(&self.perFragmentProgram, shaders: shaders)
        if successed {
            self.perfragmentUniformsLocation = self.unifromLocation(inProgram: self.perFragmentProgram)
            
            let matrixUniformBlockIndex = glGetUniformBlockIndex(self.perFragmentProgram, "transformMatrixs")
            glUniformBlockBinding(self.perFragmentProgram, matrixUniformBlockIndex, 0)
        }
        return successed
    }
    
    private func preparePerVertexProgram() -> Bool {
        let shaders = ["perVertexVS" : GLenum(GL_VERTEX_SHADER), "perVertexFS" : GLenum(GL_FRAGMENT_SHADER)]
        let successed = self.prepareGLProgram(&self.perVertexProgram, shaders: shaders)
        if successed {
            self.perVertexUnifromsLocation = self.unifromLocation(inProgram: self.perVertexProgram)
            
            let matrixUniformBlockIndex = glGetUniformBlockIndex(self.perVertexProgram, "transformMatrixs")
            glUniformBlockBinding(self.perVertexProgram, matrixUniformBlockIndex, 0)
        }
        return successed
    }
    
    private func unifromLocation(inProgram program: GLuint) -> UniformLocation {
        let ambientColorLocation = glGetUniformLocation(program, "ambient")
        let diffuseAlbedoLocation = glGetUniformLocation(program, "diffuseAlbedo")
        let specularAlbedoLocation = glGetUniformLocation(program, "specularAlbedo")
        let specularPowerLocation = glGetUniformLocation(program, "specularPower")
        let uniformLocation = UniformLocation(ambientColorLocation: ambientColorLocation, diffuseAlbedoLocation: diffuseAlbedoLocation, specularAlbedoLocation: specularAlbedoLocation, specularPowerLocation: specularPowerLocation)
        return uniformLocation
    }
    
    private func prepareGLProgram(_ glProgram: inout GLuint, shaders: [String : GLenum]) -> Bool {
        if shaders.count == 0 {
            print("No available shader.")
            return false
        }
        
        // 1. 创建GL程序
        glProgram = glCreateProgram()

        for (_, shaderInfo) in shaders.enumerated() {
            // 2. 获取着色器的源码
            guard let shaderPath = Bundle.main.path(forResource: shaderInfo.key, ofType: "glsl") else {
                print("Can not find shader file at" + shaderInfo.key)
                return false
            }
            
            // 2. 创建着色器
            let shader = glCreateShader(shaderInfo.value)
            
            // 3. 编译着色器
            if !self.compileShader(shader, filePath: shaderPath) {
                glDeleteShader(shader)
                return false
            }

            // 4. 将着色器附着至OpenGL程序
            glAttachShader(glProgram, shader)
            glDeleteShader(shader)
        }
        
        // 5. 编译GL程序
        if !self.linkProgram(program: glProgram) {
            print("Failed to link program")
            glDeleteProgram(glProgram)
            glProgram = 0
            return false
        }
        return true
    }
        
    private func compileShader(_ shader: GLuint, filePath: String) -> Bool {
        // 1. 为着色器填充源码
        var shaderSource: UnsafePointer<GLchar>?
        do {
            shaderSource = try NSString(contentsOfFile: filePath, encoding: String.Encoding.utf8.rawValue).utf8String
        } catch {
            print("Error when compile shader")
        }
        glShaderSource(shader, 1, &shaderSource, nil)
        // 2. 编译着色器
        glCompileShader(shader)
        
        // 3. 查询着色器编译错误日志
        var status: GLint = 0
        glGetShaderiv(shader, GLenum(GL_COMPILE_STATUS), &status)
        if status == 0 {
            var logLength: GLint = 0
            glGetShaderiv(shader, GLenum(GL_INFO_LOG_LENGTH), &logLength)
            var infoLogChars = [GLchar](repeating: 0, count: Int(logLength))
            glGetShaderInfoLog(shader, logLength, nil, &infoLogChars)
            let infoLogString = String(utf8String: infoLogChars) ?? ""
            print("Compile Shader Failed at \n" + filePath + "\nWith Log\n" + infoLogString)
            return false
        }
        return true
    }

    func linkProgram(program: GLuint) -> Bool {
        glLinkProgram(program)
        var status: GLint = 0
        glGetProgramiv(program, GLenum(GL_LINK_STATUS), &status)
        if status == 0 {
            var logLength: GLint = 0
            glGetProgramiv(program, GLenum(GL_INFO_LOG_LENGTH), &logLength)
            var logChars = [GLchar](repeating: 0, count: Int(logLength))
            glGetProgramInfoLog(program, GLsizei(logLength), nil, &logChars)
            let logString = String(utf8String: logChars) ?? ""
            print("Compile program failed with log \n" + logString)
            return false
        }
        
        return true
    }
            
    override func draw(_ dirtyRect: NSRect) {
        // 1. 清空颜色缓存和深度缓存
        let grayColor:[GLfloat] = [0.1, 0.1, 0.1, 0]
        glClearBufferfv(GLenum(GL_COLOR), 0, grayColor)
        var defaultDepth:GLfloat = 1;
        glClearBufferfv(GLenum(GL_DEPTH), 0, &defaultDepth)

        // 2. 设置OpenGL程序
        if self.per_vertex {
            glUseProgram(self.perVertexProgram)
        } else {
            glUseProgram(self.perFragmentProgram)
        }

        let viewPosition = GLKVector3(v: (0, 0, 20))
        let viewMatrix = GLKMatrix4MakeLookAt(viewPosition.x, viewPosition.y, viewPosition.z,
                                              0, 0, 0,
                                              0, 1, 0)

        if self.many_objects {
            for j in 0...6 {
                for i in 0...6 {
                    // 3. 填充放射矩阵缓存数据
                    glBindBufferBase(GLenum(GL_UNIFORM_BUFFER), 0, self.uniformsBuffer)
                    guard let blockAddress = glMapBufferRange(GLenum(GL_UNIFORM_BUFFER), 0, MemoryLayout<UniformsBlock>.size, GLbitfield(GL_MAP_WRITE_BIT | GL_MAP_INVALIDATE_BUFFER_BIT)) else { return }
                    let blockTypeAddress = blockAddress.bindMemory(to: UniformsBlock.self, capacity: 1)

                    let modelMatrix = GLKMatrix4MakeTranslation(Float(i) * 2.75 - 8.25, 6.75 - Float(j) * 2.25, 0)
                    let modelViewMatrix = GLKMatrix4Multiply(viewMatrix, modelMatrix)
                    let projectionMatrix = GLKMatrix4MakePerspective(GLKMathDegreesToRadians(50),
                                                                     Float(NSWidth(self.bounds) / NSHeight(self.bounds)),
                                                                     0.1,
                                                                     1000)
                    let uniformBlocl = UniformsBlock(viewMatrix: viewMatrix, modelViewMatrix: modelViewMatrix, projectionMatrix: projectionMatrix)
                    blockTypeAddress.pointee = uniformBlocl
                    glUnmapBuffer(GLenum(GL_UNIFORM_BUFFER))

                    // 4. 填充材质属性数据
                    var specularAlbedo = [GLfloat](repeating: (Float(i) + 1) / 9, count: 3)
                    let specularPower = GLfloat(powf(2, Float(j) + 2))
                    if self.per_vertex {
                        glUniform3fv(self.perVertexUnifromsLocation.specularAlbedoLocation, 1, &specularAlbedo)
                        glUniform1f(self.perVertexUnifromsLocation.specularPowerLocation, specularPower)
                    } else {
                        glUniform3fv(self.perfragmentUniformsLocation.specularAlbedoLocation, 1, &specularAlbedo)
                        glUniform1f(self.perfragmentUniformsLocation.specularPowerLocation, specularPower)
                    }

                    // 5. 渲染
                    ModelManager.shareManager.render()
                    
                    // 6. 复原OpenGL上下文相关绑定点状态
                    glBindBufferBase(GLenum(GL_UNIFORM_BUFFER), 0, 0)
                }
            }
        } else {
            // 3. 填充放射矩阵缓存数据
            glBindBufferBase(GLenum(GL_UNIFORM_BUFFER), 0, self.uniformsBuffer)
            guard let blockAddress = glMapBufferRange(GLenum(GL_UNIFORM_BUFFER), 0, MemoryLayout<UniformsBlock>.stride, GLbitfield(GL_MAP_WRITE_BIT | GL_MAP_INVALIDATE_BUFFER_BIT)) else { return }
            let blockTypeAddress = blockAddress.bindMemory(to: UniformsBlock.self, capacity: 1)
            
            let modelMatrix = GLKMatrix4MakeScale(7, 7, 7)
            let modelViewMatrix = GLKMatrix4Multiply(viewMatrix, modelMatrix)
            let projectionMatrix = GLKMatrix4MakePerspective(GLKMathDegreesToRadians(50),
                                                             Float(NSWidth(self.bounds) / NSHeight(self.bounds)),
                                                             0.1,
                                                             1000)
            let unifromBlock = UniformsBlock(viewMatrix: viewMatrix, modelViewMatrix: modelViewMatrix, projectionMatrix: projectionMatrix)
            blockTypeAddress.pointee = unifromBlock
            glUnmapBuffer(GLenum(GL_UNIFORM_BUFFER))

            // 4. 填充材质属性数据
            var specularAlbedo: [GLfloat] = [1, 1, 1]
            let specularPower: Float = 30
            if self.per_vertex {
                glUniform3fv(self.perVertexUnifromsLocation.specularAlbedoLocation, 1, &specularAlbedo)
                glUniform1f(self.perVertexUnifromsLocation.specularPowerLocation, specularPower)
            } else {
                glUniform3fv(self.perfragmentUniformsLocation.specularAlbedoLocation, 1, &specularAlbedo)
                glUniform1f(self.perfragmentUniformsLocation.specularPowerLocation, specularPower)
            }

            // 5. 渲染
            ModelManager.shareManager.render()
            
            // 6. 复原OpenGL上下文相关绑定点状态
            glBindBufferBase(GLenum(GL_UNIFORM_BUFFER), 0, 0)
        }
        glFlush()
    }
}
